# Introdução
Esse exercício referece ao que foi visto nas aulas de coleções. A ideia aqui é permitir que os alunos possam apresentar conclusões sobre o resultado obtido. Os dados aqui utilizados serão gerados de maneira pseudoaleatórios, portanto cada execução terá um resultado único
> obs: Execute em ordem para não haver problemas   
> links de referências:  
> [Coleções em python](https://phylos.net/2021-03-08/python-sequencias-e-colecoes#:~:text=Cole%C3%A7%C3%B5es%20s%C3%A3o%20conteineres%20de%20dados.%20No%20m%C3%B3dulo%20b%C3%A1sico,n%C3%A3o%20ordenada%20e%20n%C3%A3o%20indexada%2C%20sem%20membros%20duplicados.) 

## Perguntas de pesquisa e implementação

> (Crie uma nova célula de texto ou escreva na célula de código abaixo)

### Quais são as diferenças entre tupla e lista? Implemente um exemplo de cada um e apresente seus itens 

In [ ]:
# uma lista pode abordar diversos tipos de dados, desde textos até numeros
# as listas são ordenadas de valores e o seus valores podem ser repetidos mais 
# de uma vez

lista = ["banana", "maça", "larana", "pera", "banana"]
print(lista)
print(type(lista))

# já as TUPLAS funcionam da mesma forma que as listas, porém as tuplas apresentam
# valores imutáveis, ou seja, que não podem ser alterados ou re

tupla = (1,2,3,4,5)
print(tupla)
print(type(tupla))


['banana', 'maça', 'larana', 'pera', 'banana']
<class 'list'>
(1, 2, 3, 4, 5)
<class 'tuple'>


### Quais sãos as diferenças entre lista e dicionários em python? Implemente um exemplo simples de cada e apresente seus itens

In [ ]:
# as listas podem ter os seus dados já registrados pelo usuário e consulta-los
# já os dicionarios podem armazenar diversos dados em uma determinada indice (posição de dados)


# uma lista simples

lista1 = [1,2,3,4,5]
print(lista1)
print(type(lista1))

# uma dicionario simples

dicionario = ['banana', 'maça', 'larana', 'pera', 'limao']
print(dicionario[0:4])
print(type(dicionario))


[1, 2, 3, 4, 5]
<class 'list'>
['banana', 'maça', 'larana', 'pera']
<class 'list'>


### Em python o que significa quando um objeto é mutavel e imutavel? Quais são suas diferenças? Implemente ao menos um de cada na célula abaixo

In [ ]:
# a diferença entre objetos mutaveis e não mutaveis é que os objetos 
# MUTAVEIS podem ser modificados após a sua declaração de variável, já
# os IMUTAVEIS não podem ser modificados uma vez que já foram declarados

# exemplo de um objeto mutaveis
listas = []
print(type(listas))

# exemplo de um objeto imutavel
tupla = ()
print(type(tupla))

<class 'list'>
<class 'tuple'>


## Exercício prático
Uma empresa conhecida no ramo de aplicativos de inteligência artificial (IA) conhecida como **ConfIA+** gostaria de conhcer o perfil do novo público de um aplicativo lançado recentemente. O cliente acredita que através da idade e salários fornecidos (sem nomes por questões de LGPD) poderá ser feito o mapeamento do perfil e tomar decisões estratégicas. Para isso, chegou até a sua equipe uma série de operações a serem realizadas e por fim entregarmos uma conclusão para apresentar para o cliente.

## Criando os dados que serão utilizados para a atividade (Mock)

### Dados obtidos do cliente. A lista abaixo possui uma tupla com a idade e a somatória dos gastos em um determinado serviço  

In [ ]:
# SÓ EXECUTAR A CÉLULA, NÃO PRECISA ALTERAR
# Importa biblioteca para usar números aleatórios
import random as numero_aleatorio # a palavra 'as' indica que queremos dar um apelido a essa biblioteca
# Defini o tamanho da população
quantidade_pessoas = numero_aleatorio.randint(1000, 5000)

Aqui será feito o preenchimento da lista com idade e gasto por idade (SÓ EXECUTAR)

In [ ]:
# SÓ EXECUTAR A CELULA, NÃO PRECISA ALTERAR
# Preenchimento da lista dos dados com números aleatórios. 
lista = []
for indice in range(quantidade_pessoas):
  idade = numero_aleatorio.randint(18, 40) # Gera valores entre 18 e 40
  salario = numero_aleatorio.randint(2800, 10000) # Gera valores entre 2800 até 10000
  lista.append((idade, salario))
print(lista)

[(33, 7509), (25, 3744), (19, 7999), (23, 4822), (34, 7073), (37, 5416), (36, 3777), (22, 6704), (28, 9253), (30, 4318), (36, 5544), (21, 4356), (36, 8113), (20, 5646), (21, 4787), (31, 8343), (23, 5389), (36, 8097), (38, 9191), (30, 2970), (36, 2951), (29, 8228), (23, 4588), (30, 3956), (23, 7802), (31, 3899), (34, 9468), (20, 4557), (19, 6987), (37, 3362), (39, 3136), (27, 5378), (37, 5369), (32, 3417), (22, 8403), (33, 9430), (28, 9219), (32, 5188), (22, 6559), (29, 9399), (21, 6990), (18, 8120), (38, 7173), (37, 7688), (34, 5284), (22, 3916), (40, 4395), (35, 6971), (22, 2847), (30, 7133), (36, 2983), (25, 6859), (34, 7863), (37, 3265), (20, 9207), (23, 4174), (27, 3167), (40, 6444), (33, 8210), (32, 8299), (21, 8029), (27, 5562), (34, 8332), (21, 6369), (25, 6140), (39, 9245), (35, 8288), (18, 8611), (25, 7974), (19, 5159), (22, 4978), (18, 8804), (34, 6744), (39, 6567), (29, 7287), (21, 6382), (26, 9317), (19, 6061), (28, 6580), (29, 3693), (27, 6581), (25, 4531), (30, 8347), (34

## Atividade prática 

### 1) Apresente o tipo e a quantidade de itens da variavel lista declarada e atribuida acima

O cliente quer saber quantas pessoas estão usando o seu produto, para isso, apresente a quantidade de itens presente na lista e também o tipo da lista como um incremento

In [ ]:
# Preencha com sua resposta
print(len(lista))
print(type(lista))

2648
<class 'list'>


### 2) Apresente a primeiro item e o último item da lista criada acima. Cada item dessa lista é uma tupla

O cliente é uma pessoa desconfiada e preocupado com o seu negócio. Para tentar diminuir a confiança dele ele teve acesso ao primero e último item da lista para confirmar com a lista que temos basta apresentar o primeiro e o último item dessa lista. 

In [ ]:
# Preencha com sua resposta
print(lista[0])
print(lista[-1])

(19, 9649)
(18, 3441)


### 3) Apresente a soma dos cinco primeiros salários da lista criada acima

O cliente ainda está desconfiado, então ele sugeriu que você e sua equipe apresentasse a soma dos **cinco primeiros salários** da lista obtida por ele, assim ele poderá confirmar os dados que você possui e o que ele possui

In [ ]:
# Preencha com sua resposta


sum = 0
for indiceLista in range(1,6):
  sum += lista[indiceLista][1]
  print(lista[indiceLista])

print(f'A soma dos 5 primeiros salarios são {sum}')

(37, 6776)
(33, 9814)
(40, 5514)
(34, 2956)
(37, 3609)
A soma dos 5 primeiros salarios são 28669


### 4) Apresente a soma dos cinco últimos salários da lista acima
Mais uma prova de conceito que o cliente quer fazer. Ele solicitou agora a soma dos últimos cinco salarios da lista criada acima

In [ ]:
# Preencha com sua resposta
sum = 0
for indiceLista in range(1,6):
  sum += lista[-indiceLista][1]
  print(lista[indiceLista])

print(f'A soma dos 5 ultimos salários são {sum}')

(33, 9814)
(40, 5514)
(34, 2956)
(37, 3609)
(28, 4641)
A soma dos 5 primeiros salarios são 38075


### 5) Faça a média das idades entre a posição 25 até 155 e apresente a média de seus salários em uma linha

Agora o grupo de executivos do cliente solicitou uma tarefa um pouco estranha, porém faz sentido em estatística. Ele quer pegar uma amostra dessa população (lista) criada verificar se ela está bem distribuida. Para isso ele pediram para realizar a média das idades e salário das posições 25 até 155.

In [ ]:
# Preecha com sua resposta

print(lista[25])
print(lista[155])

mde = 0 

for i in range (25,155):
  mde += lista[i][1]
print(lista[0])
print(lista[0][0])
print(f'A média de salários entre 25 até 255 é {mde}')




(40, 8064)
(31, 8137)
(29, 7325)
29
A média de salários entre 25 até 255 é 855232


### 6) Dado a lista acima, faça os seguintes cálculos: (Obrigatório realizar com laço de repetição, sem usar funções de soma ou ajuda de bibliotecas externas)

Faça a somatória das idades e realize a média com duas casas decimais arredondadas
Faça a somatória dos gastos e realize a média com duas casas decimais truncadas

In [ ]:
# Média de idades
sum = 0 

for i in range (0,5):
  sum += lista[i][0]
print(lista[0])
print(lista[0][0])
print(f'A  das idades é {sum}')

#########################################
soma = 0

for item in lista:
  idade - item[0]
  soma +- idade
  print(f"a média é {int(soma/len(lista)))}")



(37, 6776)
37
A soma das idades é 181


In [ ]:
# Média dos salários

soma = 0
print(len(lista))

for i in range (0,len(lista)):
  soma += lista[i][1]


print(f'soma do salário é {soma/len(lista)}')

################################################

soma = 0 
for item in lista:
  salario = item[1]
  soma +- salario
  


2876
soma do salário é 6428.302155771906


### **CONCLUSÃO**

nesse levantamento, vimos que existem diversas diferenças em certos tipos de dados e como eles podem ser classificados,
como no exemplo apresentado, foram feitas diversas técnicas para que possamos 
chegar dentro do resultado esperado, essas que serviram para 

- verificar a posição e os tipos dos dados levantados
- apresentar a soma dos salarios dos funcionários
- caucular a média de idade dos funcionários 

diante as operações feitas e o que foi apresentado, chegamos na conclusão de que:

- a SOMA dos cinco primeiros e dos cinco ultimos salários dos funcionários estão entre RS 28,000.00 a RS 38,000.00

- a MÉDIA salárial dos funcionários com a idade entre 25 e 255 é R$ 85,523.20

- a somatória das idades apresentadas são 181 e a somatória dos gastos (salário)
são de 6428.302155771906

assim, possamos ter o claro conhecimento dos gastos da empresa, juntamente com a faixa de idade dos colaboradores